src/behavior.cpp

```diff
@@ -0,0 +1,90 @@
+#include "behavior.h"
+
+using namespace behavior;
+
+namespace behavior
+{
+sequential default_sequential;
+fallback default_fallback;
+}
+
+std::pair<status, const node *> sequential::evaluate( const Creature *subject,
+        const std::vector<const node *> children ) const
+{
+    for( const node *child : children ) {
+        std::pair<status, const node *> outcome = child->tick( subject );
+        switch( outcome.first ) {
+            case running:
+            case failure:
+                return outcome;
+            case success:
+                ;
+        }
+    }
+    return std::make_pair( success, nullptr );
+}
+
+std::pair<status, const node *> fallback::evaluate( const Creature *subject,
+        const std::vector<const node *> children ) const
+{
+    for( const node *child : children ) {
+        std::pair<status, const node *> outcome = child->tick( subject );
+        switch( outcome.first ) {
+            case running:
+            case success:
+                return outcome;
+            case failure:
+                ;
+        }
+    }
+    return std::make_pair( failure, nullptr );
+}
+
+void node::set_strategy( const strategy *new_strategy )
+{
+    _strategy = new_strategy;
+}
+void node::set_predicate( std::function<status( const Creature *subject )> new_predicate )
+{
+    _predicate = new_predicate;
+}
+void node::set_goal( const std::string &new_goal )
+{
+    _goal = new_goal;
+}
+void node::add_child( const node *new_child )
+{
+    _children.push_back( new_child );
+}
+
+std::pair<enum status, const node *> node::tick( const Creature *subject ) const
```

Comment:
For those of us in the cheap seats, a comment here explaining what is going on would be helpful.  I think I can puzzle it out from the tests but `tick` and `evaluate` seem to be about 90% of the core code so a little more explanation would be great.

The current AI code suffers a great deal from being untestable and inscrutable.  You're fixing one of the mistakes by designing this with testing in mind, please add enough docs to make limit the other.

Reply:
Let me know if there's anything that needs more clarity.